### ISSUE TITLE
automerge-c: Nice to haves for automerge-go

### ISSUE THREAD

**Issue by ConradIrwin** [automerge/automerge#451]
I spent a bit of time trying to make automerge-go, using automerge-c as a basis. I was able to get reading/writing values, and loading/saving working (after #450), but there are a few things that would have made the task easier. When you're ready to focus on the C API, a few things "nice to haves" that came to mind.

* Go doesn’t support union types well, so if AMvalue was an opaque struct and had accessor methods for each type that would have been better (though I’m not sure how to handle someone calling the wrong accessor on the wrong variant; maybe just don’t do that :D). For go I ended up generating a bunch of stub methods.
* When iterating over a map or a list, you can either get a Value or an ObjId. It’d be tidier if the objId was also returned by AMmapItemValue instead of having a separate accessor there (how this is handled in AMmapGet).
* When iterating over a map, if you get an item with an object id, you need to retain the AMresult that contains all the mapItems just to keep the object id pointer valid. It’d be nice if objectid was returned by value instead of reference (but not sure the performance impact of that). 
* (This is also true for methods like AMmapGet that do combine the Value and ObjId into one struct: because all other values can be converted to Go types immediately, the only time that I need to retain the AMresult from AMmapGet is if it returned an object id, which makes calling AMfree at the right time a bit more complicated than I’d like.)
* If you write an int > 2**52, the javascript version crashes on load; it’d be good to either return an error, or fix the js implementation.
* (not sure if a bug or a feature?) AMactorIdBytes() allows creating a 0-length actor id.
* It took me a while to get the build working to generate libautomerge.a and automerge.h (running on Mac, the scripts had a few syntax errors according to my local copy of bash; and I am missing a lot of implicit dependencies). It'd be nice if there was a place I could just download the built versions.
* (not sure if a real problem?) Currently automerge.h contains some structs that say they depend on architecture; it'd be nice (but not essential) if there was one header file that worked across multiple architectures. This may already be the case in practice, but I've only tested on darwin/arm64.

I'm not very familiar with rust, (nor that familiar with C), so there may be reasons for the above choices that I don't understand; and as I said, I was able to build what I wanted so none of this should be considered a bug.

**Comment by ConradIrwin** [comment]
A few more things here as I've made more progress:

* What kind of concurrency guarantees are there? Is it safe to call any functions concurrently, or do I need to mediate all access with a mutex? I am hoping to create a scenario where one thread is listening on the network and applying changes, while others are reading/writing the doc directly. I'm not sure how rust works in this regard, but my ideal would be that the library is thread-safe and can be called heedlessly – but would be satisfied to know what the requirements are so I can add muteness myself. (This may be a can of worms, do MapItems continue to be well behaved in the case where you're concurrently updating the map...)
* It'd be nice to document whether a function that returns an AMresult can actually error if the pre-conditions are met. Because go has no exceptions, I'd like to avoid providing an error return if it is not possible for a call to error.
* (not sure if a bug or a feature?). Calling `AMlistPut(doc, objId, C.SIZE_MAX, false, v)` works to append to an empty list; but fails to append to a longer list. It took me a while to figure out why append worked sometimes (in either case I should have been using `true`, but it took me longer to debug than I'd care to admit :D).

**Comment by jkankiewicz** [comment]
> What kind of concurrency guarantees are there?

@ConradIrwin, an `AMdoc` is not thread-safe because Automerge began its life as a pure JavaScript library and the default Rust implementation of the document class that it exposes is based upon the assumption that there's only one document per thread. A thread-safe version of the document class can be instantiated within a Rust application but that's not the version exposed by the C API because it's unlikely that a multi-threaded application would be written purely in C.

> [D]o MapItems continue to be well behaved in the case where you're concurrently updating the map...)

@ConradIrwin, yes, any iterator continues to be well-behaved as you're concurrently updating its underlying container object because it's actually referencing a static copy of that object's elements stored within the `AMresult` that it was created from. This also enables one to create as many iterators as one needs, both forwards and backwards, from the same `AMresult` without fear that they'll return different values for the same index.
The caveat is, of course, that if you've been updating the document that the `AMresult` was returned from then the data that it contains may be stale.

**Comment by ConradIrwin** [comment]
@jkankiewicz Good to know! Does that mean that it *is* safe to call any method that doesn't take a "receiver" concurrently? (e.g. things like `AMcreate()`) or does the library have shared mutable state outside of a document?

**Comment by jkankiewicz** [comment]
> [D]oes the library have shared mutable state outside of a document?

@ConradIrwin, each `AMdoc` is self-contained; there are no implicit linkages between any two of them. When synchronizing two documents, the current state of their synchronization is solely maintained within an `AMsyncState` external to both of them because neither one "knows" about the other.

**Comment by ConradIrwin** [comment]
Great, thanks for confirming!

**Comment by jkankiewicz** [comment]
> Go doesn’t support union types well, so if AMvalue was an opaque struct and had accessor methods for each type that would have been better (though I’m not sure how to handle someone calling the wrong accessor on the wrong variant; maybe just don’t do that :D). For go I ended up generating a bunch of stub methods.

@ConradIrwin, as you've demonstrated, it's not necessary for `AMvalue` to become an opaque type  in order for the contents of an instance of it to be retrieved using an accessory function.
It makes sense to furnish an additional header containing such functions so that they won't have to be written by everyone seeking to author bindings based upon automerge-c.

**Comment by jkankiewicz** [comment]
> When iterating over a map or a list, you can either get a Value or an ObjId. It’d be tidier if the objId was also returned by AMmapItemValue instead of having a separate accessor there (how this is handled in AMmapGet).

@ConradIrwin, are you asking for `AMlistGet()` and `AMmapGet()` to always embed an `AMobjId const*` in addition to the `AMvalue` within the `AMresult`?

Doing so would also entail `AMresultValue()` and `AMobjItemsNext()` returning an `AMitem` struct consisting of directly-accessible `AMobjId const*` and  `AMvalue` members, similar to the `AMobjItem` struct but without requiring `AMobjItemValue()` and `AMobjItemObjId()` calls to retrieve its contents.
It follows that the `AMlistItem` and `AMmapItem` structs would then consist of an `AMitem` member and a `size_t` (index) or `AMbyteSpan` (key) member, respectively.

Do I have that right?

**Comment by ConradIrwin** [comment]
@jkankiewicz i need to push up the latest code, and will try to do so tomorrow so you can explore it.

With respect to AMvalue being an opaque struct, the reason I’d like this is so that I don’t have to have 3 sets of such functions. In the current implementation I have accessor functions that take an AMresult and return AMresultValue(r).tag (or whatever); I also need a set that take an AMmapItem and an AMlistItem if I started using the values from iteration. As each C function needs a Go wrapper this is a comparatively large amount of extra code (both in the source and when compiled).

I did it this way because if I try and access a C union from Go it gets returned as an array of bytes. An alternative I could possibly use is a set of wrapper function that takes an array of bytes and casts them back to the Union on the C side; but that felt quite fragile.

For object ids I think an objIdClone function would solve the problem - it’s not a huge problem in any case; because I can always reference count, but it seemed quite inefficient to retain the AMresult containing all map items just to keep a single objId alive.

The other solution to the iterator ownership would be for the AMmapItem to contain just the key and change hashes, and have a function AMmapItemValue() that returns an AMresult with and AMvalue - just as though you’d called AMmapGet(). 

I’m not sure what the overall performance difference would be like; but it would also solve the oddity of having a different function for getting the value and the objectId, and let me re-use the go logic for converting the return value of AMmapGet when accessing mapItems.

**Comment by ConradIrwin** [comment]
I've pushed up the latest code so you can see what I'm doing now; suggestions for improvements welcome!

The autogenerated accessors are here: https://github.com/ConradIrwin/automerge-go/blob/main/cgo.go#L17-L37, and the approach to getting the value for map iteration using just Get so I can re-use the code that handles AMmapGet is here: https://github.com/ConradIrwin/automerge-go/blob/main/cgo.go#L867-L870

The other thing I'd like to do is add support for linux/x86_64, but I haven't put the effort in to configure cross-compilation from my Mac yet!

**Comment by ConradIrwin** [comment]
@jkankiewicz one other question about the API:

Currently `AMcommit` sometimes returns a change hash, and sometimes returns void. In the case that it returns void, you need to call `AMemptyChange` to create a commit.

I can obviously work around this, but it's a bit awkward because the abstractions I created assume that the return type of functions is consistent between calls (and void is represented as a different type to change hashes).

I'd prefer two alternatives:
* (my preference) `AMcommit` always returns a change hash - and internally it creates an empty change if necessary
* (less ideal, but still ok) `AMcommit` returns either a change hash, or returns an error explaining why.

Do you know why there's a distinction between the two in the original API?

**Comment by alexjg** [comment]
The distinction is because in the case that there are no operations to commit then no change is created and consequently there is no hash to return. The return type in Rust is `Option<ChangeHash>`

**Comment by ConradIrwin** [comment]
Makes sense. 
It probably is an error to commit when you have created no operations; so it seems reasonable to have a separate API for doing that if you really need to.


Would you be open to a PR to change it (either automerge-rs or just automerge-c) to return an error to explain why the change wasn't created instead of an option?

**Comment by alexjg** [comment]
I think in the Rust API returning an `Option<ChangeHash>` already communicates quite well what is going on - `Option` is intended for this kind of thing. I have no opinions about the C API :sweat_smile: so I'll leave that to you and @jkankiewicz  :slightly_smiling_face: .

**Comment by ConradIrwin** [comment]
Makes sense. As the C API already returns a result, it seems better to make it Result<ChangeHash, Error> instead of Result<Option<ChangeHash>, Error>

**Comment by jkankiewicz** [comment]
@ConradIrwin, I hope to have addressed most, if not all, of the issues that you've raised through the changes that are captured in PR #530; please let me know to what extent I've succeeded.

> Go doesn’t support union types well, so if AMvalue was an opaque struct and had accessor methods for each type that would have been better...

There are no union types anymore: you get an `AMitem` struct's value by passing its `AMitem*` to `AMitemTo<to-type>()` and you set a value by calling `AMitemFrom<from-type>()`, `AM<to-type>From<from-type>()`, `AMlistPut<from-type>()`, or `AMmapPut<from-type>()`.
What do you think about replacing all of the  `AMlistPut<from-type>()`s with a single `AMlistPut()` that takes an `AMitem*` argument and replacing all of the `AMmapPut<from-type>()`s with a single `AMmapPut()` that takes an `AMitem*` argument? I can non-destructively add `AMlistPut()` and `AMmapPut()` for you to experiment with if you're interested.

>  ...(though I’m not sure how to handle someone calling the wrong accessor on the wrong variant; maybe just don’t do that :D). ...

For a simple <to-type>, wrongly calling `AMitemTo<to-type>()` returns a default value and for a complex <to-type>, wrongly calling `AMitemTo<to-type>()` returns `NULL`; in either case, the specific value returned by `AMitemTo<to-type>()` in this situation is documented in its post condition. It behooves you to pass its `AMitem*` to `AMitemValType()` first in order to determine which `AMitemTo<to-type>()` to call. 

> ...For go I ended up generating a bunch of stub methods.

I hope that those stub methods can be removed now.

> When iterating over a map or a list, you can either get a Value or an ObjId. It’d be tidier if the objId was also returned by AMmapItemValue instead of having a separate accessor there (how this is handled in AMmapGet).

An `AMitem` struct contains an object ID in addition to a value: you get one's `AMobjId*` by passing its `AMitem*` to `AMitemObjId()`.

> When iterating over a map, if you get an item with an object id, you need to retain the AMresult that contains all the mapItems just to keep the object id pointer valid. It’d be nice if objectid was returned by value instead of reference (but not sure the performance impact of that).

You can create a new `AMresult` struct for an `AMitem` struct by passing its `AMitem*` to `AMitemResult()` and then you only have to retain the new `AMresult*` for the `AMitem` struct; the `AMresult` struct for the map can be safely freed.

> (This is also true for methods like AMmapGet that do combine the Value and ObjId into one struct: because all other values can be converted to Go types immediately, the only time that I need to retain the AMresult from AMmapGet is if it returned an object id, which makes calling AMfree at the right time a bit more complicated than I’d like.)

An `AMitem` struct contains both a value and an object ID so when you wish to retain only the value, and it can be converted into a Go type immediately, you can free the `AMitem` struct's corresponding `AMresult` struct immediately but when you wish to retain only the object ID then you'll need to also retain the `AMitem` struct's corresponding `AMresult` struct. Passing an `AMitem` struct's `AMitem*` to `AMitemResult()` will produce a new `AMresult` struct specific to it.

> If you write an int > 2**52, the javascript version crashes on load; it’d be good to either return an error, or fix the js implementation.

@alexjg, are you already aware of this?

> (not sure if a bug or a feature?) AMactorIdBytes() allows creating a 0-length actor id.

@alexjg, is `automerge::ActorId::from()` supposed to allow the creation of a zero-length actor ID?

> It took me a while to get the build working to generate libautomerge.a and automerge.h (running on Mac, the scripts had a few syntax errors according to my local copy of bash; and I am missing a lot of implicit dependencies). It'd be nice if there was a place I could just download the built versions.

Hopefully a Homebrew package will be added for automerge-c some day. I chose CMake for building the library so that it would be fairly easy to create a package for it on any platform. CMake does output a relevant error message when a package dependency like Cargo, cmocka or Doxygen hasn't been installed yet. CMake can already find an installed Homebrew cmocka package on a Mac with Intel silicon but it needs additional help on a Mac with Apple silicon. @heckj opened [a ticket](https://github.com/Homebrew/homebrew-core/issues/108287) to have this addressed but the Homebrew maintainers don't seem to be convinced that it's a problem with their package.

> (not sure if a real problem?) Currently automerge.h contains some structs that say they depend on architecture; it'd be nice (but not essential) if there was one header file that worked across multiple architectures. This may already be the case in practice, but I've only tested on darwin/arm64.

The implementation of the `AMitems` struct contains a pointer and it's assumed that CMake is building automerge-c for the platform that it's running on. I can make it possible to override the implicit pointer size by passing an additional command line option to CMake.

**Comment by ConradIrwin** [comment]
We’ve made most of these changes
